Prevent pods from defaulting to zero second grace periods

smarterclayton opened the pull request:
A zero second grace period on pods is a special value and is intended for exceptional use (to break the safety guarantees when a node is partitioned). When spec.terminationGracePeriodSeconds (which overrides the default of 30s) is zero, pods are instantly deleted without waiting for the kubelet to ensure the process is not running. The intent of this default was not to allow users to bypass that process (since the normal behavior of a pod is not to bypass kubelet shutdown) and the documentation was written as such. Users who wish to bypass the kubelet's participation must set gracePeriod to zero on the delete options call.

With this change the value 1 second is substituted when users specify 0, resulting in the kubelet participating in the deletion process and only a minor delay in practice.

The controlling design for pod safety is https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/pod-safety.md and I believe we simply missed this during the early review and no one ever blew themselves up with this footgun. We did not intend to offer the footgun as a feature of defaulting as described here in the doc:

> A user may specify an interval on the pod called the termination grace period that defines the minimum amount of time the pod will have to complete the termination phase, and all components will honor this interval.

That matches the value we placed in the field godoc:

> The grace period is the duration in seconds after the processes running in the pod are sent a termination signal and the time when the processes are forcibly halted with a kill signal.

The intent was the default is the minimum, that the kubelet would be involved in deletion, and that gracePeriod=0 on a delete operation was the only way a user would be able to perform force delete. Unfortunately we didn't actually handle 0 specially at the time we implemented this and it has remained ambiguous since then.

As this changes the observed behavior of the system (prioritizing safety and the documented behavior over the unsafe behavior), this is also an API change. Kube is consistent by default, and bypassing the kubelet is both a safety (accidental consistency loss) and operational hazard (deleting a large number of pods without the kubelet means that resources are still allotted to those workers and could take an arbitrary amount of time to clean up).

Discovered while doing a review of clusters in the wild - some workloads were setting this thinking that it simply meant "as fast as possible".

/kind bug
/kind api-change
/sig node

```release-note
The pod `spec.terminationGracePeriodSeconds` field unintentionally allowed a zero value to be copied directly to the pod during deletion, causing the pod to bypass kubelet shutdown during deletion (aka force deletion). This was unintentional - force deletion was only meant to be possible on an explicit delete with the gracePeriod field provided to delete options.  Pods with `spec.terminationGracePeriodSeconds=0` now are interpreted the same as `spec.terminationGracePeriodSeconds=1` and will no longer be deleted without the kubelet having a chance to clean up the pod.  The kubelet will shut the pod down as fast as it can, and force deleting a pod is now only possible by invoking delete or delete collection on pods with the gracePeriod=0 field.  
```

```docs
```

smarterclayton committed:
Prevent pods from defaulting to zero second grace periods

The zero grace period is a special value and is intended for
exceptional use (to break the safety guarantees when a node
is partitioned). When spec.terminationGracePeriodSeconds is zero,
pods are instantly deleted without waiting for the kubelet to
ensure the process is not running. The intent of this default
was not to allow users to bypass that process (since the normal
behavior of a pod is not to violate kubelet shutdown) and the
documentation was written as such. Users who wish to bypass
the kubelet's participation must set gracePeriod to zero on
the delete options call.

The value one is substituted when users specify 0, resulting
in the kubelet participating in the deletion process and only
a minor delay.

smarterclayton committed:
cleanup: Fix typos in the pod strategy test

smarterclayton commented:
/assign @liggitt 
/assign @derekwaynecarr

smarterclayton commented:
/priority important-soon

smarterclayton commented:
Wow, we have e2e tests that are racing on the same pod name: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102025/pull-kubernetes-e2e-kind-ipv6/1393342213345251328

```


Kubernetes e2e suite: [sig-network] Services should implement service.kubernetes.io/service-proxy-name expand_less | 1m31s
-- | --
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service.go:1883 May 14 23:32:30.493: Unexpected error:     <*errors.StatusError \| 0xc0008e59a0>: {         ErrStatus: {             TypeMeta: {Kind: "", APIVersion: ""},             ListMeta: {                 SelfLink: "",                 ResourceVersion: "",                 Continue: "",                 RemainingItemCount: nil,             },             Status: "Failure",             Message: "object is being deleted: pods \"verify-service-down-host-exec-pod\" already exists",             Reason: "AlreadyExists",             Details: {                 Name: "verify-service-down-host-exec-pod",                 Group: "",                 Kind: "pods",                 UID: "",                 Causes: nil,                 RetryAfterSeconds: 0,             },             Code: 409,         },     }     object is being deleted: pods "verify-service-down-host-exec-pod" already exists occurred /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service.go:2705
```

These tests were force deleting pods, then starting a pod of the same name and execing to it - which wouldn't necessarily guarantee that an exec session was even going to the new pod (at least, historically it wouldn't).  I will fix the e2e tests that assume that deletion like this is safe (it isn't)

smarterclayton committed:
e2e tests were reusing the same name AND force deleting pods

Previously these tests were relying on force deletion, which allowed
new pods with the same name to be created before the pods on the
nodes completed. This meant that a test that did:

1. create pod foo/bar
2. exec pod foo/bar "run some test"
3. delete pod foo/bar (force)
4. create pod foo/bar
5. exec pod foo/bar "some other test"

could end up execing into the terminating container if the delete
notification was delayed on watch.

smarterclayton committed:
integration: Simulate Kubelet deletion behavior as necessary

The kubelet owns termination of scheduled pods. In integration
tests, run pod deletion on pods that have requested termination
as necessary.

dims commented:
/hold

for reviews and prevent accidental merging :)

smarterclayton commented:
`Kubernetes e2e suite: [sig-api-machinery] Garbage collector should keep the rc around until all its pods are deleted if the deleteOptions says so ` in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102025/pull-kubernetes-e2e-kind-ipv6/1393391901620572160 is broken because it's taking over 1m30s minutes to delete 10 pods that have single second termination grace period.  That's a very long time - it points to either kubelet teardown issues (several of which exist and I suspect there may be containerd issues not diagnosed) or another issue with qps.  Looking at the pod remaining on the first worker:

```
May 15 02:42:53 kind-worker kubelet[251]: I0515 02:42:53.451499     251 kubelet.go:1948] "SyncLoop DELETE" source="api" pods=[gc-4860/simpletest.rc-2mlnf]
...
May 15 02:43:11 kind-worker kubelet[251]: I0515 02:43:11.814953     251 kubelet_pods.go:967] "Pod is terminated, but some containers are still running" pod="gc-4860/simpletest.rc-2mlnf"
...
May 15 02:44:16 kind-worker kubelet[251]: I0515 02:44:16.382949     251 kuberuntime_container.go:714] "Killing container with a grace period override" pod="gc-4860/simpletest.rc-2mlnf" podUID=ae1207d1-5e23-4844-aff1-8e5a6ee2b160 containerName="nginx" containerID="containerd://5aa9bb1fe965bfc14eb5f8ddeab3f0a22362c795b4b9ebba3656d8d9a91a0d57" gracePeriod=2
...
# this takes FAR too long - what is going on
...
May 15 02:44:16 kind-worker kubelet[251]: I0515 02:44:16.383385     251 event.go:291] "Event occurred" object="gc-4860/simpletest.rc-2mlnf" kind="Pod" apiVersion="v1" type="Normal" reason="Killing" message="Stopping container nginx"
May 15 02:44:16 kind-worker kubelet[251]: I0515 02:44:16.532080     251 kuberuntime_container.go:722] "Container exited normally" pod="gc-4860/simpletest.rc-2mlnf" podUID=ae1207d1-5e23-4844-aff1-8e5a6ee2b160 containerName="nginx" containerID="containerd://5aa9bb1fe965bfc14eb5f8ddeab3f0a22362c795b4b9ebba3656d8d9a91a0d57"
```

This implies we have a kubelet bug where some events are being lost / the worker is saturated, or a call is failing in containerd.  Needs investigation.  This is yet another reason why force delete in e2e hides SLO failures in underlying components - I would expect subsecond container kills from time the delete is received, even if cleanup happens.

Something very wrong / inadequate is happening in code + environment.  @rphillips can you help me find an owner to run down what would cause the kubelet to get this bogged down (might be container runtime, hard to say).

smarterclayton commented:
/retest

dims commented:
/retest

smarterclayton commented:
Integration tests were not emulating the Kubelet behavior in a few cases (daemonsets, some scheduling).   Since it's fairly cheap to perform the emulation I updated the integration tests with that minimal config.  Still have to look at why ingress is failing.

smarterclayton commented:
In https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102025/pull-kubernetes-conformance-kind-ga-only-parallel/1394433436277542912

```
May 17 23:43:30 kind-worker kubelet[250]: I0517 23:43:30.318925     250 kubelet.go:1949] "SyncLoop DELETE" source="api" pods=[gc-4943/simpletest-rc-to-be-deleted-kmvj9]
...
May 17 23:43:30 kind-worker kubelet[250]: I0517 23:43:30.690732     250 kubelet.go:1971] "SyncLoop (PLEG): event for pod" pod="gc-4943/simpletest-rc-to-be-deleted-kmvj9" event=&{ID:ba114bd4-67bf-4a5c-8169-3dbba7319673 Type:ContainerStarted Data:3c6a282b0267628899f7b9fad1e49ca5869bf0b521f91d787301b4865c2fe2e9}
May 17 23:43:30 kind-worker kubelet[250]: I0517 23:43:30.690765     250 kubelet.go:2047] "Pod has completed execution and should be deleted from the API server" pod="gc-4943/simpletest-rc-to-be-deleted-kmvj9" syncType="sync"
...
May 17 23:43:37 kind-worker kubelet[250]: I0517 23:43:37.375646     250 kubelet_pods.go:967] "Pod is terminated, but some containers are still running" pod="gc-4943/simpletest-rc-to-be-deleted-kmvj9"
...
May 17 23:44:33 kind-worker kubelet[250]: I0517 23:44:33.937765     250 kubelet_pods.go:967] "Pod is terminated, but some containers are still running" pod="gc-4943/simpletest-rc-to-be-deleted-kmvj9"
...
May 17 23:44:42 kind-worker kubelet[250]: I0517 23:44:42.702530     250 kuberuntime_container.go:938] "Removing container" containerID="3c6a282b0267628899f7b9fad1e49ca5869bf0b521f91d787301b4865c2fe2e9"
May 17 23:44:42 kind-worker kubelet[250]: E0517 23:44:42.703136     250 remote_runtime.go:334] "ContainerStatus from runtime service failed" err="rpc error: code = NotFound desc = an error occurred when try to find container \"3c6a282b0267628899f7b9fad1e49ca5869bf0b521f91d787301b4865c2fe2e9\": not found" containerID="3c6a282b0267628899f7b9fad1e49ca5869bf0b521f91d787301b4865c2fe2e9"
...
May 17 23:44:56 kind-worker kubelet[250]: I0517 23:44:56.843564     250 status_manager.go:614] "Pod fully terminated and removed from etcd" pod="gc-4943/simpletest-rc-to-be-deleted-kmvj9"
```

smarterclayton commented:
pull-kubernetes-e2e-ubuntu-gce-network-policies  is failing because v1beta1 ingress is now off, https://github.com/kubernetes/kubernetes/pull/102030 is what is fixing that.

ehashman commented:
/triage accepted
/assign

smarterclayton commented:
https://github.com/kubernetes/kubernetes/pull/98742 is probably related to why this is cropping up now, but the root issue is that dispatchWork is not the right place to touch container status from PLEG events, so we need fix the fundamental race condition on deleted pods by ensuring the worker is the one who does all summarization and cleanup.

smarterclayton commented:
Opened https://github.com/kubernetes/kubernetes/issues/102149 to cover the issue, debugging it here.

smarterclayton commented:
Exposed some novel container runtime bugs:

```
pod pod-submit-status-2-6 on node kind-worker2 container unexpected exit code 128: start=1970-01-01 00:00:00 +0000 UTC end=2021-05-19 18:08:25 +0000 UTC reason=StartError message=failed to create containerd task: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:402: getting the final child's pid from pipe caused: EOF: unknown
pod pod-submit-status-2-14 on node kind-worker container unexpected exit code 128: start=1970-01-01 00:00:00 +0000 UTC end=2021-05-19 18:09:47 +0000 UTC reason=StartError message=failed to create containerd task: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:402: getting the final child's pid from pipe caused: EOF: unknown
```

Those should likely be suppressed, or are race conditions

wzshiming commented:
It looks like this PR #98507 fixes the same problem

aojea commented on pkg/registry/core/pod/strategy_test.go:
```diff
@@ -1600,3 +1599,128 @@ func TestDropNonEphemeralContainerUpdates(t *testing.T) {
+			name: "terminal (succeeded) pod is deleted immediately",
+			obj: &core.Pod{
+				Spec:   core.PodSpec{NodeName: "test", TerminationGracePeriodSeconds: int64p(1)},
+				Status: core.PodStatus{Phase: core.PodFailed},
```
you mean `core.PodSucceeded` here, right?

smarterclayton replied:
Yeah cut and paste error

aojea commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -145,6 +145,9 @@ func (podStrategy) CheckGracefulDelete(ctx context.Context, obj runtime.Object,
 		// use the default value if set, or deletes the pod immediately (0)
```
should we fix this comment to explain that if value 0 it will use 1?

msau42 replied:
Also update the comment for `pod.Spec.TerminationGracePeriodSeconds` in types.go

aojea commented on test/e2e/network/service.go:
```diff
@@ -2697,11 +2698,21 @@ func createPodOrFail(f *framework.Framework, ns, name string, labels map[string]
+		if err != nil && apierrors.IsAlreadyExists(err) && strings.Contains(err.Error(), "is being deleted") {
+			return false, nil
+		}
+		return true, err
```
we use to retry on all errors, ie. "I can't connect to apiserver" or "apiserver returns 500 because etcd ... " or "connection timeout"
Maybe just logging a message only for those specific errors you are looking for ?

smarterclayton replied:
Not sure what you mean?  This retries *only* alreadyexists, and reports the same error (return true, err ensures that if an error happens it goes to the next line)

aojea replied:
maybe I'm not understanding it well, but I think that if there is an error, that is not "alreadyExists", this will fail the test.

aojea replied:
ahhhh ok, nice, I got it, we were trying to do this in another test.
Scratch my comment

smarterclayton commented:
https://github.com/kubernetes/kubernetes/pull/98507#issuecomment-845197777 connects the two issues - that PR doesn't go far enough because there are bugs in other loops (loops are confused between "fully terminated" and "terminating", and the pod worker needs to run long enough to ensure the pod is finally deleted).

I think there are also changes needed to syncPod - i'm unconvinced shutdown is correct because today podWorker stops too early.

wzshiming commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -145,6 +145,9 @@ func (podStrategy) CheckGracefulDelete(ctx context.Context, obj runtime.Object,
 		// use the default value if set, or deletes the pod immediately (0)
 		if pod.Spec.TerminationGracePeriodSeconds != nil {
 			period = *pod.Spec.TerminationGracePeriodSeconds
+			if period == 0 {
```
```suggestion
			if period < 1 {
```
This value may be negative, maybe it's better to deal with it together

smarterclayton replied:
Validation should prevent that.

wzshiming replied:
But this verification is missing #98866

Random-Liu commented:
/assign @Random-Liu

smarterclayton commented:
> The scenario described here was discussed and warned at SIG Node several times long time ago, and one particular example I were using is disk space clean up after forcing deletion, assuming Kubelet would follow up and execute soon. I forgot the counter argument provided back then. But at the end we made the compromise and allow 0 for pod in the api, so that apiserver can remove the pod object directly (force delete)

Note that force delete via a delete call is still allowed - this change blocks a workload (statefulset, deployment, job) or pod from being able to specify that force deletion.  You mentioned on the call that "fast deletion is desirable" and that in the future we want to "move cleanup from blocking deletion to being a cleanup process" - I think that's roughly compatible with the change I was going for here:

1. Make teardown of containers more deterministic (reduces tail latency)
2. Make termination be able to interrupt syncPod via context cancellation (needed to be able to reduce grace period anyway)
3. Optimize status reporting so that the final termination happens predictably (requires both changes like this and status manager)
4. Remove container cleanup from being a prereq for etcd deletion because we want to keep logs until deletion (required for docker compat)

The combined impact of all of those changes (only one of which would not be in this effort - status optimization) should get deletion time down significantly.  I *think* with the race conditions fixed here, we'll be able to shut down very fast with fairly minimal impact to "post deletion load" on kubelet.

One source of latency this doesn't address is volume unmount, which is slow, but probably needs to happen before we can delete the pod no matter what (otherwise someone using pod name uniqueness to prevent mulitple pods from accessing the same NFS share could potentially be broken).

smarterclayton commented:
Now that this is getting close to complete, moving the pod worker part to #102344 - please take all review there.

smarterclayton committed:
e2e: Pod test should wait a bit longer for delete

Now that pods are not force deleted, some tests that wait for pods
to be deleted take a realistic amount of time, which is about a
minute from deletion request to kubelet finally being able to
delete the pod from etcd due to latency in the status pipeline
in dense e2e tests.

Increase the timeout of this test to two minutes.

aojea commented:
/retest
100-performance ssh errors

aojea commented:
:eyes: two consecutive failures on the gce-100-performance job doesn't look good, will rerun but we should check this is not regressing performance
/retest

the verify job fails because
```
 Please run hack/update-openapi-spec.sh
```

aojea commented:
I'm worried that this is going to affect current e2e tests that "abuse" this zero second grace period,
```
grep -r --color=never TerminationGracePeriodSeconds test | grep -E '0|zero'                                                revert-101604-tuning-grpc 
test/integration/daemonset/daemonset_test.go:                                   TerminationGracePeriodSeconds: &zero,
test/integration/daemonset/daemonset_test.go:           TerminationGracePeriodSeconds: &zero,
test/integration/deployment/deployment_test.go: if *allOldRs[0].Spec.Template.Spec.TerminationGracePeriodSeconds == newTGPS {
test/integration/cronjob/cronjob_test.go:                                                       TerminationGracePeriodSeconds: &zero64,
test/integration/etcd/data.go:                  Stub:             `{"metadata":{"name":"crs3"},"data":{"name":"abc","namespace":"default","creationTimestamp":null,"Spec":{"Replicas":0,"Selector":{"matchLabels":{"foo":"bar"}},"Template":{"creationTimestamp":null,"labels":{"foo":"bar"},"Spec":{"Volumes":null,"InitContainers":null,"Containers":null,"RestartPolicy":"Always","TerminationGracePeriodSeconds":null,"ActiveDeadlineSeconds":null,"DNSPolicy":"ClusterFirst","NodeSelector":null,"ServiceAccountName":"","AutomountServiceAccountToken":null,"NodeName":"","SecurityContext":null,"ImagePullSecrets":null,"Hostname":"","Subdomain":"","Affinity":null,"SchedulerName":"","Tolerations":null,"HostAliases":null}},"VolumeClaimTemplates":null,"ServiceName":""},"Status":{"ObservedGeneration":null,"Replicas":0}},"revision":0}`,
test/utils/runners.go:  if config.TerminationGracePeriodSeconds == nil || *config.TerminationGracePeriodSeconds < 0 {
grep: test/images/agnhost/agnhost: binary file matches
test/e2e/node/pods.go:                  framework.ExpectNotEqual(lastPod.Spec.TerminationGracePeriodSeconds, 0)
test/e2e/apimachinery/namespace.go:     maxWaitSeconds := int64(60) + *pod.Spec.TerminationGracePeriodSeconds
test/e2e/apimachinery/resource_quota.go:                                        TerminationGracePeriodSeconds: &zero,
test/e2e/apimachinery/webhook.go:                                       TerminationGracePeriodSeconds: &zero,
test/e2e/apimachinery/aggregator.go:                                    TerminationGracePeriodSeconds: &zero,
test/e2e/apimachinery/garbage_collector.go:                     TerminationGracePeriodSeconds: &zero,
test/e2e/apimachinery/garbage_collector.go:                                                     TerminationGracePeriodSeconds: &zero,
test/e2e/apimachinery/crd_conversion_webhook.go:                                        TerminationGracePeriodSeconds: &zero,
test/e2e/framework/deployment/fixtures.go:                                      TerminationGracePeriodSeconds: &zero,
test/e2e/framework/deployment/fixtures.go:                                      TerminationGracePeriodSeconds: &zero,
test/e2e/apps/replica_set.go:                                   TerminationGracePeriodSeconds: &zero,
test/e2e/apps/replica_set.go:                                   "TerminationGracePeriodSeconds": &zero,
test/e2e/apps/deployment.go:    d.Spec.Template.Spec.TerminationGracePeriodSeconds = &zero
test/e2e/apps/rc.go:                                    TerminationGracePeriodSeconds: &zero,
test/e2e/auth/service_accounts.go:                              TerminationGracePeriodSeconds: &zero,
test/e2e/network/netpol/model.go:                       TerminationGracePeriodSeconds: &zero,
```

 and we'll end with a lot of flakes because of that ... 
should we delay this to the start of the 1.23 cycle and change the tests that assume the behavior accordenly?

smarterclayton commented:
This PR is *fixing* the bugs in the e2e tests.  It was the fact that we were abusing this in tests, and then that a large set of flaky issues in the Kubelet were being completely ignored, that led to the argument that this behavior is unacceptable in production usage (Kube is not designed to bypass Kubelet termination).  This PR led to #102344, so it was actually finding the problems that we were completely ignoring.  Remember that force delete doesn't stop things on the kubelet faster - instead, it loads the kubelet up in a way that admins / the rest of the system can't observe (except via metrics) - so the performance regression is measured latency (we weren't actually measuring the impact of e2e on the cluster).

All that being said, yes this should be targeted for 1.23 because it's a change to the behavior of the system and we will need time to go find all the issues that force deletion was hiding.

smarterclayton commented:
/retest

smarterclayton commented:
/test all

aojea commented:
@smarterclayton it needs some tuning `hack/update-openapi-spec.sh` 
 
> All that being said, yes this should be targeted for 1.23 because it's a change to the behavior of the system and we will need time to go find all the issues that force deletion was hiding.

I think we should merge this as soon as possible

smarterclayton commented:
Talking with jordan he suggested three things:

1. a warning
2. a feature gate to control the change
3. a metric

all of which are reasonable as we are changing the behavior of the system and it might impact someone creating a pod and then deleting and expect immediate deletion (due to the implicit zero).  I'm going to draft a KEP which clarifies the behavior and brings in pod safety rules (still the original guiding element).

smarterclayton commented:
Hrm, there's a very interesting bug in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102025/pull-kubernetes-e2e-kind-ipv6/1498726712261742592:

```
Kubernetes  e2e suite: [sig-network] Proxy version v1 A set of valid responses are  returned for both pod and service Proxy [Conformance] expand_less | 1m2s
-- | --
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:633 Mar  1 18:56:13.682: Pod didn't start within time out period Unexpected error:     <*errors.errorString \| 0xc000210250>: {         s: "timed out waiting for the condition",     }     timed out waiting for the condition occurred /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/proxy.go:408
```

The status update for agnhost is delivered to the kubelet (we detect that container agnhost started and thus transition the phase to running) but then the kubelet logs show that the status update was "unchanged" at 18:55:34.494020 and then at 18:55:56.088086 we sent a patch that shows the container in waiting.  I *think* this is head of line blocking that #107897 fixes (we put status updates in a queue and read them in order, but we really only need the most recent status).  If so, unrelated to this PR.

```
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.324498     262 kubelet.go:2077] "SyncLoop (PLEG): event for pod" pod="proxy-4151/agnhost" event=&{ID:d54c3888-9dd2-410f-9713-0b94cfc44bc0 Type:ContainerStarted Data:15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3}
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325199     262 kubelet.go:1488] "syncPod enter" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325214     262 kubelet_pods.go:1429] "Generating pod status" pod="proxy-4151/agnhost"
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325243     262 kubelet_pods.go:1441] "Got phase for pod" pod="proxy-4151/agnhost" oldPhase=Pending phase=Running
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325565     262 volume_manager.go:404] "Waiting for volumes to attach and mount for pod" pod="proxy-4151/agnhost"
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325630     262 volume_manager.go:435] "All volumes are attached and mounted for pod" pod="proxy-4151/agnhost"
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325767     262 kuberuntime_manager.go:712] "computePodActions got for pod" podActions={KillPod:false CreateSandbox:false SandboxID:d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1 Attempt:0 NextInitContainerToStart:nil ContainersToStart:[] ContainersToKill:map[] EphemeralContainersToStart:[]} pod="proxy-4151/agnhost"
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325899     262 kubelet.go:1712] "syncPod exit" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325929     262 pod_workers.go:979] "Processing pod event done" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 updateType=0
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.325941     262 pod_workers.go:888] "Processing pod event" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 updateType=0
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.413079     262 desired_state_of_world_populator.go:317] "Added volume to desired state" pod="proxy-4151/agnhost" volumeName="kube-api-access-nvjrs" volumeSpecName="kube-api-access-nvjrs"
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.426072     262 reconciler.go:247] "Starting operationExecutor.MountVolume for volume \"kube-api-access-nvjrs\" (UniqueName: \"kubernetes.io/projected/d54c3888-9dd2-410f-9713-0b94cfc44bc0-kube-api-access-nvjrs\") pod \"agnhost\" (UID: \"d54c3888-9dd2-410f-9713-0b94cfc44bc0\") Volume is already mounted to pod, but remount was requested." pod="proxy-4151/agnhost"
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.426826     262 projected.go:183] Setting up volume kube-api-access-nvjrs for pod d54c3888-9dd2-410f-9713-0b94cfc44bc0 at /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volumes/kubernetes.io~projected/kube-api-access-nvjrs
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.427050     262 atomic_writer.go:161] pod proxy-4151/agnhost volume kube-api-access-nvjrs: no update required for target directory /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volumes/kubernetes.io~projected/kube-api-access-nvjrs
Mar 01 18:55:33 kind-worker2 kubelet[262]: I0301 18:55:33.427081     262 operation_generator.go:739] "MountVolume.SetUp succeeded for volume \"kube-api-access-nvjrs\" (UniqueName: \"kubernetes.io/projected/d54c3888-9dd2-410f-9713-0b94cfc44bc0-kube-api-access-nvjrs\") pod \"agnhost\" (UID: \"d54c3888-9dd2-410f-9713-0b94cfc44bc0\") " pod="proxy-4151/agnhost"
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.493569     262 kubelet.go:1488] "syncPod enter" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.493588     262 kubelet_pods.go:1429] "Generating pod status" pod="proxy-4151/agnhost"
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.493640     262 kubelet_pods.go:1441] "Got phase for pod" pod="proxy-4151/agnhost" oldPhase=Running phase=Running
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.494020     262 status_manager.go:475] "Ignoring same status for pod" pod="proxy-4151/agnhost" status={Phase:Running Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:13 +0000 UTC Reason: Message:} {Type:Ready Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:33 +0000 UTC Reason: Message:} {Type:ContainersReady Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:33 +0000 UTC Reason: Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:13 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:fc00:f853:ccd:e793::3 PodIP:fd00:10:244:1::f6 PodIPs:[{IP:fd00:10:244:1::f6}] StartTime:2022-03-01 18:55:13 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:agnhost State:{Waiting:nil Running:&ContainerStateRunning{StartedAt:2022-03-01 18:55:33 +0000 UTC,} Terminated:nil} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:true RestartCount:0 Image:k8s.gcr.io/e2e-test-images/agnhost:2.33 ImageID:k8s.gcr.io/e2e-test-images/agnhost@sha256:5b3a9f1c71c09c00649d8374224642ff7029ce91a721ec9132e6ed45fa73fd43 ContainerID:containerd://15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3 Started:0xc0034c6899}] QOSClass:BestEffort EphemeralContainerStatuses:[]}
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.494256     262 volume_manager.go:404] "Waiting for volumes to attach and mount for pod" pod="proxy-4151/agnhost"
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.494358     262 volume_manager.go:435] "All volumes are attached and mounted for pod" pod="proxy-4151/agnhost"
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.494522     262 kuberuntime_manager.go:712] "computePodActions got for pod" podActions={KillPod:false CreateSandbox:false SandboxID:d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1 Attempt:0 NextInitContainerToStart:nil ContainersToStart:[] ContainersToKill:map[] EphemeralContainersToStart:[]} pod="proxy-4151/agnhost"
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.494649     262 kubelet.go:1712] "syncPod exit" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.494694     262 pod_workers.go:979] "Processing pod event done" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 updateType=0
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.545019     262 desired_state_of_world_populator.go:317] "Added volume to desired state" pod="proxy-4151/agnhost" volumeName="kube-api-access-nvjrs" volumeSpecName="kube-api-access-nvjrs"
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.644813     262 reconciler.go:247] "Starting operationExecutor.MountVolume for volume \"kube-api-access-nvjrs\" (UniqueName: \"kubernetes.io/projected/d54c3888-9dd2-410f-9713-0b94cfc44bc0-kube-api-access-nvjrs\") pod \"agnhost\" (UID: \"d54c3888-9dd2-410f-9713-0b94cfc44bc0\") Volume is already mounted to pod, but remount was requested." pod="proxy-4151/agnhost"
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.645072     262 projected.go:183] Setting up volume kube-api-access-nvjrs for pod d54c3888-9dd2-410f-9713-0b94cfc44bc0 at /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volumes/kubernetes.io~projected/kube-api-access-nvjrs
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.645372     262 atomic_writer.go:161] pod proxy-4151/agnhost volume kube-api-access-nvjrs: no update required for target directory /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volumes/kubernetes.io~projected/kube-api-access-nvjrs
Mar 01 18:55:34 kind-worker2 kubelet[262]: I0301 18:55:34.645409     262 operation_generator.go:739] "MountVolume.SetUp succeeded for volume \"kube-api-access-nvjrs\" (UniqueName: \"kubernetes.io/projected/d54c3888-9dd2-410f-9713-0b94cfc44bc0-kube-api-access-nvjrs\") pod \"agnhost\" (UID: \"d54c3888-9dd2-410f-9713-0b94cfc44bc0\") " pod="proxy-4151/agnhost"
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.113992     262 factory.go:258] Using factory "containerd" for container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3"
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.115400     262 manager.go:988] Added container: "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3" (aliases: [15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3 /docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3], namespace: "containerd")
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.115691     262 handler.go:325] Added event &{/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3 2022-03-01 18:55:32.949377956 +0000 UTC containerCreation {<nil>}}
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.115744     262 container.go:531] Start housekeeping for container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3"
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.143145     262 factory.go:258] Using factory "containerd" for container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1"
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.144665     262 manager.go:988] Added container: "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1" (aliases: [d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1 /docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1], namespace: "containerd")
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.144930     262 handler.go:325] Added event &{/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1 2022-03-01 18:55:28.305058425 +0000 UTC containerCreation {<nil>}}
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.145005     262 container.go:531] Start housekeeping for container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1"
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.166532     262 factory.go:262] Factory "containerd" was unable to handle container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0"
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.166539     262 factory.go:262] Factory "systemd" was unable to handle container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0"
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.166550     262 factory.go:255] Factory "raw" can handle container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0", but ignoring.
Mar 01 18:55:49 kind-worker2 kubelet[262]: I0301 18:55:49.166558     262 manager.go:925] ignoring container "/docker/9a00b8cc23eec20d1be7279e14bf116ccfc1b327bdbc59425d24cfdcd2fac1ca/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0"
Mar 01 18:55:55 kind-worker2 kubelet[262]: I0301 18:55:55.866672     262 request.go:519] Waited for 192.688646ms due to client-side throttling, not priority and fairness, request: GET:https://kind-control-plane:6443/api/v1/namespaces/proxy-4151/pods/agnhost
Mar 01 18:55:56 kind-worker2 kubelet[262]: I0301 18:55:56.066988     262 request.go:519] Waited for 197.368267ms due to client-side throttling, not priority and fairness, request: PATCH:https://kind-control-plane:6443/api/v1/namespaces/proxy-4151/pods/agnhost/status
Mar 01 18:55:56 kind-worker2 kubelet[262]: I0301 18:55:56.088086     262 status_manager.go:624] "Patch status for pod" pod="proxy-4151/agnhost" patch="{\"metadata\":{\"uid\":\"d54c3888-9dd2-410f-9713-0b94cfc44bc0\"},\"status\":{\"$setElementOrder/conditions\":[{\"type\":\"Initialized\"},{\"type\":\"Ready\"},{\"type\":\"ContainersReady\"},{\"type\":\"PodScheduled\"}],\"conditions\":[{\"lastProbeTime\":null,\"lastTransitionTime\":\"2022-03-01T18:55:13Z\",\"status\":\"True\",\"type\":\"Initialized\"},{\"lastProbeTime\":null,\"lastTransitionTime\":\"2022-03-01T18:55:13Z\",\"message\":\"containers with unready status: [agnhost]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"Ready\"},{\"lastProbeTime\":null,\"lastTransitionTime\":\"2022-03-01T18:55:13Z\",\"message\":\"containers with unready status: [agnhost]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"ContainersReady\"}],\"containerStatuses\":[{\"image\":\"k8s.gcr.io/e2e-test-images/agnhost:2.33\",\"imageID\":\"\",\"lastState\":{},\"name\":\"agnhost\",\"ready\":false,\"restartCount\":0,\"started\":false,\"state\":{\"waiting\":{\"reason\":\"ContainerCreating\"}}}],\"hostIP\":\"fc00:f853:ccd:e793::3\",\"startTime\":\"2022-03-01T18:55:13Z\"}}"
Mar 01 18:55:56 kind-worker2 kubelet[262]: I0301 18:55:56.088171     262 status_manager.go:633] "Status for pod updated successfully" pod="proxy-4151/agnhost" statusVersion=1 status={Phase:Pending Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:13 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:13 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [agnhost]} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:13 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [agnhost]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2022-03-01 18:55:13 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:fc00:f853:ccd:e793::3 PodIP: PodIPs:[] StartTime:2022-03-01 18:55:13 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:agnhost State:{Waiting:&ContainerStateWaiting{Reason:ContainerCreating,Message:,} Running:nil Terminated:nil} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:k8s.gcr.io/e2e-test-images/agnhost:2.33 ImageID: ContainerID: Started:0xc00124d029}] QOSClass:BestEffort EphemeralContainerStatuses:[]}
Mar 01 18:55:56 kind-worker2 kubelet[262]: I0301 18:55:56.089865     262 kubelet.go:2052] "SyncLoop RECONCILE" source="api" pods=[proxy-4151/agnhost]
Mar 01 18:56:13 kind-worker2 kubelet[262]: I0301 18:56:13.697818     262 logs.go:320] "Finished parsing log file" path="/var/log/pods/proxy-4151_agnhost_d54c3888-9dd2-410f-9713-0b94cfc44bc0/agnhost/0.log"
Mar 01 18:56:13 kind-worker2 kubelet[262]: I0301 18:56:13.697880     262 httplog.go:131] "HTTP" verb="GET" URI="/containerLogs/proxy-4151/agnhost/agnhost?tailLines=42" latency="992.758µs" userAgent="Go-http-client/1.1" audit-ID="" srcIP="[fc00:f853:ccd:e793::4]:32930" resp=200
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123558     262 kubelet.go:2055] "SyncLoop DELETE" source="api" pods=[proxy-4151/agnhost]
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123608     262 pod_workers.go:625] "Pod is marked for graceful deletion, begin teardown" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123649     262 pod_workers.go:888] "Processing pod event" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 updateType=1
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123666     262 pod_workers.go:996] "Pod worker has observed request to terminate" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123696     262 kubelet.go:1719] "syncTerminatingPod enter" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123709     262 kubelet_pods.go:1429] "Generating pod status" pod="proxy-4151/agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123758     262 kubelet_pods.go:1441] "Got phase for pod" pod="proxy-4151/agnhost" oldPhase=Running phase=Running
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123848     262 kubelet.go:1749] "Pod terminating with grace period" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 gracePeriod=30
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123923     262 kuberuntime_container.go:718] "Killing container with a grace period override" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containerName="agnhost" containerID="containerd://15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3" gracePeriod=30
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.123943     262 kuberuntime_container.go:722] "Killing container with a grace period" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containerName="agnhost" containerID="containerd://15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3" gracePeriod=30
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.124019     262 event.go:294] "Event occurred" object="proxy-4151/agnhost" fieldPath="spec.containers{agnhost}" kind="Pod" apiVersion="v1" type="Normal" reason="Killing" message="Stopping container agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: E0301 18:56:20.125880     262 event.go:267] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"agnhost.16d856b08f64c0b0", GenerateName:"", Namespace:"proxy-4151", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"Pod", Namespace:"proxy-4151", Name:"agnhost", UID:"d54c3888-9dd2-410f-9713-0b94cfc44bc0", APIVersion:"v1", ResourceVersion:"17216", FieldPath:"spec.containers{agnhost}"}, Reason:"Killing", Message:"Stopping container agnhost", Source:v1.EventSource{Component:"kubelet", Host:"kind-worker2"}, FirstTimestamp:time.Date(2022, time.March, 1, 18, 56, 20, 123893936, time.Local), LastTimestamp:time.Date(2022, time.March, 1, 18, 56, 20, 123893936, time.Local), Count:1, Type:"Normal", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'events "agnhost.16d856b08f64c0b0" is forbidden: unable to create new content in namespace proxy-4151 because it is being terminated' (will not retry!)
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.161257     262 manager.go:1044] Destroyed container: "/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3" (aliases: [15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3 /kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3], namespace: "containerd")
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.161302     262 handler.go:325] Added event &{/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3 2022-03-01 18:56:20.16129489 +0000 UTC m=+511.863183509 containerDeletion {<nil>}}
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.217023     262 kuberuntime_container.go:731] "Container exited normally" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containerName="agnhost" containerID="containerd://15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.283025     262 manager.go:1044] Destroyed container: "/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1" (aliases: [d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1 /kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1], namespace: "containerd")
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.283072     262 handler.go:325] Added event &{/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0/d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1 2022-03-01 18:56:20.283064772 +0000 UTC m=+511.984953383 containerDeletion {<nil>}}
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.643025     262 kuberuntime_manager.go:1035] "getSandboxIDByPodUID got sandbox IDs for pod" podSandboxID=[d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1] pod="proxy-4151/agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.644313     262 kubelet.go:1791] "Post-termination container state" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containers=[{Name:agnhost State:exited ExitCode:2 FinishedAt:2022-03-01T18:56:20.13992941Z}]
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.644348     262 kubelet.go:1798] "Pod termination stopped all running containers" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.644378     262 kubelet.go:1800] "syncTerminatingPod exit" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.644393     262 pod_workers.go:1014] "Pod terminated all containers successfully" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.644420     262 pod_workers.go:979] "Processing pod event done" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 updateType=1
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.644433     262 pod_workers.go:888] "Processing pod event" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 updateType=2
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.645631     262 desired_state_of_world_populator.go:258] "Removing volume from desired state" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 volumeName="kube-api-access-nvjrs"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.736762     262 reconciler.go:192] "operationExecutor.UnmountVolume started for volume \"kube-api-access-nvjrs\" (UniqueName: \"kubernetes.io/projected/d54c3888-9dd2-410f-9713-0b94cfc44bc0-kube-api-access-nvjrs\") pod \"d54c3888-9dd2-410f-9713-0b94cfc44bc0\" (UID: \"d54c3888-9dd2-410f-9713-0b94cfc44bc0\") "
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.736811     262 subpath_linux.go:244] Cleaning up subpath mounts for /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volume-subpaths/kube-api-access-nvjrs
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.736872     262 projected.go:366] Tearing down volume kube-api-access-nvjrs for pod d54c3888-9dd2-410f-9713-0b94cfc44bc0 at /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volumes/kubernetes.io~projected/kube-api-access-nvjrs
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.737061     262 empty_dir_linux.go:99] Statfs_t of /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volumes/kubernetes.io~projected/kube-api-access-nvjrs: {Type:16914836 Bsize:4096 Blocks:13369771 Bfree:13369768 Bavail:13369768 Files:6684885 Ffree:6684876 Fsid:{Val:[0 0]} Namelen:255 Frsize:4096 Flags:4128 Spare:[0 0 0 0]}
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.737107     262 mount_linux.go:294] Unmounting /var/lib/kubelet/pods/d54c3888-9dd2-410f-9713-0b94cfc44bc0/volumes/kubernetes.io~projected/kube-api-access-nvjrs
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.739707     262 operation_generator.go:899] UnmountVolume.TearDown succeeded for volume "kubernetes.io/projected/d54c3888-9dd2-410f-9713-0b94cfc44bc0-kube-api-access-nvjrs" (OuterVolumeSpecName: "kube-api-access-nvjrs") pod "d54c3888-9dd2-410f-9713-0b94cfc44bc0" (UID: "d54c3888-9dd2-410f-9713-0b94cfc44bc0"). InnerVolumeSpecName "kube-api-access-nvjrs". PluginName "kubernetes.io/projected", VolumeGidValue ""
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.758960     262 generic.go:155] "GenericPLEG" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containerID="15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3" oldState=running newState=exited
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.758993     262 generic.go:155] "GenericPLEG" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containerID="d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1" oldState=running newState=exited
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.759620     262 kuberuntime_manager.go:1035] "getSandboxIDByPodUID got sandbox IDs for pod" podSandboxID=[d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1] pod="proxy-4151/agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.760871     262 generic.go:418] "PLEG: Write status" pod="proxy-4151/agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.760932     262 generic.go:296] "Generic (PLEG): container finished" podID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containerID="15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3" exitCode=2
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.760950     262 kubelet.go:1808] "syncTerminatedPod enter" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.760961     262 kubelet.go:2077] "SyncLoop (PLEG): event for pod" pod="proxy-4151/agnhost" event=&{ID:d54c3888-9dd2-410f-9713-0b94cfc44bc0 Type:ContainerDied Data:15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3}
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.760965     262 kubelet_pods.go:1429] "Generating pod status" pod="proxy-4151/agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.760985     262 kubelet.go:2077] "SyncLoop (PLEG): event for pod" pod="proxy-4151/agnhost" event=&{ID:d54c3888-9dd2-410f-9713-0b94cfc44bc0 Type:ContainerDied Data:d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1}
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.760998     262 kubelet_pods.go:1441] "Got phase for pod" pod="proxy-4151/agnhost" oldPhase=Running phase=Failed
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.761044     262 volume_manager.go:444] "Waiting for volumes to unmount for pod" pod="proxy-4151/agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.761107     262 volume_manager.go:471] "All volumes are unmounted for pod" pod="proxy-4151/agnhost"
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.761115     262 kubelet.go:1821] "Pod termination unmounted volumes" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.763323     262 kubelet.go:1842] "Pod termination removed cgroups" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.763431     262 kubelet.go:1847] "Pod is terminated and will need no more status updates" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.763442     262 kubelet.go:1849] "syncTerminatedPod exit" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.763451     262 pod_workers.go:1069] "Pod is complete and the worker can now stop" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.763484     262 pod_workers.go:957] "Processing pod event done" pod="proxy-4151/agnhost" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 updateType=2
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.763337     262 manager.go:1044] Destroyed container: "/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0" (aliases: [], namespace: "")
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.763499     262 handler.go:325] Added event &{/kubelet/kubepods/besteffort/podd54c3888-9dd2-410f-9713-0b94cfc44bc0 2022-03-01 18:56:20.763493445 +0000 UTC m=+512.465382050 containerDeletion {<nil>}}
Mar 01 18:56:20 kind-worker2 kubelet[262]: I0301 18:56:20.838324     262 reconciler.go:300] "Volume detached for volume \"kube-api-access-nvjrs\" (UniqueName: \"kubernetes.io/projected/d54c3888-9dd2-410f-9713-0b94cfc44bc0-kube-api-access-nvjrs\") on node \"kind-worker2\" DevicePath \"\""
Mar 01 18:56:21 kind-worker2 kubelet[262]: I0301 18:56:21.765195     262 generic.go:155] "GenericPLEG" podUID=d54c3888-9dd2-410f-9713-0b94cfc44bc0 containerID="15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3" oldState=exited newState=non-existent
Mar 01 18:56:21 kind-worker2 kubelet[262]: I0301 18:56:21.797965     262 kuberuntime_manager.go:1035] "getSandboxIDByPodUID got sandbox IDs for pod" podSandboxID=[d8eebef8cac27edd498e15b54767774d9529278815a9b25c3730488f4642cbd1] pod="proxy-4151/agnhost"
Mar 01 18:56:21 kind-worker2 kubelet[262]: I0301 18:56:21.798925     262 generic.go:418] "PLEG: Write status" pod="proxy-4151/agnhost"
Mar 01 18:56:24 kind-worker2 kubelet[262]: I0301 18:56:24.866456     262 request.go:519] Waited for 359.411956ms due to client-side throttling, not priority and fairness, request: GET:https://kind-control-plane:6443/api/v1/namespaces/proxy-4151/pods/agnhost
Mar 01 18:56:26 kind-worker2 kubelet[262]: I0301 18:56:26.267554     262 request.go:519] Waited for 1.373835715s due to client-side throttling, not priority and fairness, request: PATCH:https://kind-control-plane:6443/api/v1/namespaces/proxy-4151/pods/agnhost/status
Mar 01 18:56:26 kind-worker2 kubelet[262]: I0301 18:56:26.300979     262 status_manager.go:624] "Patch status for pod" pod="proxy-4151/agnhost" patch="{\"metadata\":{\"uid\":\"d54c3888-9dd2-410f-9713-0b94cfc44bc0\"},\"status\":{\"$setElementOrder/conditions\":[{\"type\":\"Initialized\"},{\"type\":\"Ready\"},{\"type\":\"ContainersReady\"},{\"type\":\"PodScheduled\"}],\"conditions\":[{\"lastTransitionTime\":\"2022-03-01T18:55:33Z\",\"message\":null,\"reason\":null,\"status\":\"True\",\"type\":\"Ready\"},{\"lastTransitionTime\":\"2022-03-01T18:55:33Z\",\"message\":null,\"reason\":null,\"status\":\"True\",\"type\":\"ContainersReady\"}],\"containerStatuses\":[{\"containerID\":\"containerd://15b1f5ed1483cd823b1ca6a57448b4bc2943baec17200ceadbf44a11c4078dc3\",\"image\":\"k8s.gcr.io/e2e-test-images/agnhost:2.33\",\"imageID\":\"k8s.gcr.io/e2e-test-images/agnhost@sha256:5b3a9f1c71c09c00649d8374224642ff7029ce91a721ec9132e6ed45fa73fd43\",\"lastState\":{},\"name\":\"agnhost\",\"ready\":true,\"restartCount\":0,\"started\":true,\"state\":{\"running\":{\"startedAt\":\"2022-03-01T18:55:33Z\"}}}],\"phase\":\"Running\",\"podIP\":\"fd00:10:244:1::f6\",\"podIPs\":[{\"ip\":\"fd00:10:244:1::f6\"}]}}"
```

pohly commented:
/reopen

This is still relevant. I ran into this unexpected behavior of `TerminationGracePeriodSeconds=0` recently and created a doc update to warn about it (https://github.com/kubernetes/kubernetes/pull/112564), but personally I would prefer to remove this obviously unintended footgun - i.e., let's revive this PR and merge that instead of my doc update.

pohly commented:
/reopen
/assign
